src/frequenz/sdk/timeseries/logical_meter/__init__.py

```diff
@@ -0,0 +1,12 @@
+"""A logical meter for calculating high level metrics for a microgrid.
+
+Copyright
+Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+License
+MIT
+"""
+
+from .logical_meter import LogicalMeter
```

Comment:
If you are only exposing `LogicalMeter`, I would also make the whole module `logical_meter` hidden to keep the public interface clean:

```suggestion
from ._logical_meter import LogicalMeter
```

And also the whole package, so in `timeseries/__init__.py` I would do too an import of:

```py
from ._logical_meter import LogicalMeter
```

On top of what's already there.

Another option, if we want to give users access to the formula engine would be to keep the package `timeseries.logical_meter` visible and only hide `timeseries.logical_meter._logical_meter`.

Reply:
Ah, already updated.

Reply:
`from sdk.timeseries.logical_meter import LogicalMeter` is now possible.  `from sdk.timeseries import LogicalMeter` is running into cyclic imports for some reason.

Reply:
I guess it comes from here:

Cyclic import (src.frequenz.sdk.actor -> src.frequenz.sdk.actor._data_sourcing -> src.frequenz.sdk.actor._data_sourcing.microgrid_api_source -> src.frequenz.sdk.timeseries -> src.frequenz.sdk.timeseries._logical_meter -> src.frequenz.sdk.timeseries._logical_meter._logical_meter -> src.frequenz.sdk.timeseries._logical_meter._formula_builder)

(there are more similar, I get these errors using `nox -R -s pylint`; strangely it is reported for `noxfile.py`, but I guess it's only because it is the last file in the command-line)

`_formula_builder` is importing `from ...actor import ChannelRegistry, ComponentMetricRequest`. I guess we always need to use the *leaf* nodes internally, it is a bit annoying as it will break more stuff if we move stuff internally but I guess it is fair trade-off to have a nice public interface.

So it should be `from ...actor._channel_registry import ChannelRegistry; from ..actor._microgrid_api_source import ComponentMetricRequest` and in `_microgrid_api_source` it should be `from ...timeseries._sample import Sample`.

Reply:
Another solution would be to make package dependencies a DAG instead of a graph, so there are no cyclic dependencies between packages. In this case it looks like anything that it is using `actor`s shouldn't live in `timeseries`. So the `actor` package knows about `timeseries` but `timeseries` doesn't know about the `actor` package.

For now I would go with importing leaf modules only, to avoid having to rethink the package structure again, but in the long term I think not having cycling dependencies among internal packages would be the way to go (when restructuring the `micrigrid` package, did exactly this, it was based in which sub-modules knew about which other sub-modules, and this is why `graph` didn't ended up in the `component` sub-package, even when it is a *component* graph, but it implied cyclic dependencies between `client` and `component`).

Reply:
With these changes it works:

<details>
<summary>Details</summary>


```console
 $ git status 
On branch logical-meter
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   src/frequenz/sdk/actor/_data_sourcing/microgrid_api_source.py
        modified:   src/frequenz/sdk/actor/_resampling.py
        modified:   src/frequenz/sdk/timeseries/__init__.py
        renamed:    src/frequenz/sdk/timeseries/logical_meter/__init__.py -> src/frequenz/sdk/timeseries/_logical_meter/__init__.py
        renamed:    src/frequenz/sdk/timeseries/logical_meter/_formula_builder.py -> src/frequenz/sdk/timeseries/_logical_meter/_formula_builder.py
        renamed:    src/frequenz/sdk/timeseries/logical_meter/_formula_engine.py -> src/frequenz/sdk/timeseries/_logical_meter/_formula_engine.py
        renamed:    src/frequenz/sdk/timeseries/logical_meter/_formula_steps.py -> src/frequenz/sdk/timeseries/_logical_meter/_formula_steps.py
        renamed:    src/frequenz/sdk/timeseries/logical_meter/_logical_meter.py -> src/frequenz/sdk/timeseries/_logical_meter/_logical_meter.py
        renamed:    src/frequenz/sdk/timeseries/logical_meter/_tokenizer.py -> src/frequenz/sdk/timeseries/_logical_meter/_tokenizer.py
```

```diff
diff --git a/src/frequenz/sdk/actor/_data_sourcing/microgrid_api_source.py b/src/frequenz/sdk/actor/_data_sourcing/microgrid_api_source.py
index f7fad2f..bad1895 100644
--- a/src/frequenz/sdk/actor/_data_sourcing/microgrid_api_source.py
+++ b/src/frequenz/sdk/actor/_data_sourcing/microgrid_api_source.py
@@ -20,7 +20,7 @@ from ...microgrid.component import (
     InverterData,
     MeterData,
 )
-from ...timeseries import Sample
+from ...timeseries._sample import Sample
 from .._channel_registry import ChannelRegistry
 
 
diff --git a/src/frequenz/sdk/actor/_resampling.py b/src/frequenz/sdk/actor/_resampling.py
index 604553c..2d45231 100644
--- a/src/frequenz/sdk/actor/_resampling.py
+++ b/src/frequenz/sdk/actor/_resampling.py
@@ -12,7 +12,8 @@ from typing import Dict, Sequence
 from frequenz.channels import Receiver, Sender
 from frequenz.channels.util import MergeNamed, Select, Timer
 
-from ..timeseries import GroupResampler, ResamplingFunction, Sample
+from ..timeseries._resampler import GroupResampler, ResamplingFunction
+from ..timeseries._sample import Sample
 from ._channel_registry import ChannelRegistry
 from ._data_sourcing import ComponentMetricRequest
 from ._decorator import actor
diff --git a/src/frequenz/sdk/timeseries/__init__.py b/src/frequenz/sdk/timeseries/__init__.py
index f52c7c5..a853e42 100644
--- a/src/frequenz/sdk/timeseries/__init__.py
+++ b/src/frequenz/sdk/timeseries/__init__.py
@@ -12,9 +12,11 @@ This module provides tools to operate on timeseries.
 
 from ._resampler import GroupResampler, Resampler, ResamplingFunction
 from ._sample import Sample
+from ._logical_meter import LogicalMeter
 
 __all__ = [
     "GroupResampler",
+    "LogicalMeter",
     "Resampler",
     "ResamplingFunction",
     "Sample",
diff --git a/src/frequenz/sdk/timeseries/logical_meter/__init__.py b/src/frequenz/sdk/timeseries/_logical_meter/__init__.py
similarity index 100%
rename from src/frequenz/sdk/timeseries/logical_meter/__init__.py
rename to src/frequenz/sdk/timeseries/_logical_meter/__init__.py
diff --git a/src/frequenz/sdk/timeseries/logical_meter/_formula_builder.py b/src/frequenz/sdk/timeseries/_logical_meter/_formula_builder.py
similarity index 96%
rename from src/frequenz/sdk/timeseries/logical_meter/_formula_builder.py
rename to src/frequenz/sdk/timeseries/_logical_meter/_formula_builder.py
index 7bd5c36..559b2ab 100644
--- a/src/frequenz/sdk/timeseries/logical_meter/_formula_builder.py
+++ b/src/frequenz/sdk/timeseries/_logical_meter/_formula_builder.py
@@ -6,7 +6,8 @@
 
 from frequenz.channels import Receiver, Sender
 
-from ...actor import ChannelRegistry, ComponentMetricRequest
+from ...actor._channel_registry import ChannelRegistry
+from ...actor._data_sourcing.microgrid_api_source import ComponentMetricRequest
 from ...microgrid.component import ComponentMetricId
 from .._sample import Sample
 from ._formula_engine import FormulaEngine
diff --git a/src/frequenz/sdk/timeseries/logical_meter/_formula_engine.py b/src/frequenz/sdk/timeseries/_logical_meter/_formula_engine.py
similarity index 100%
rename from src/frequenz/sdk/timeseries/logical_meter/_formula_engine.py
rename to src/frequenz/sdk/timeseries/_logical_meter/_formula_engine.py
diff --git a/src/frequenz/sdk/timeseries/logical_meter/_formula_steps.py b/src/frequenz/sdk/timeseries/_logical_meter/_formula_steps.py
similarity index 100%
rename from src/frequenz/sdk/timeseries/logical_meter/_formula_steps.py
rename to src/frequenz/sdk/timeseries/_logical_meter/_formula_steps.py
diff --git a/src/frequenz/sdk/timeseries/logical_meter/_logical_meter.py b/src/frequenz/sdk/timeseries/_logical_meter/_logical_meter.py
similarity index 96%
rename from src/frequenz/sdk/timeseries/logical_meter/_logical_meter.py
rename to src/frequenz/sdk/timeseries/_logical_meter/_logical_meter.py
index ede2a43..8fd160f 100644
--- a/src/frequenz/sdk/timeseries/logical_meter/_logical_meter.py
+++ b/src/frequenz/sdk/timeseries/_logical_meter/_logical_meter.py
@@ -9,7 +9,8 @@ from typing import Dict, List
 
 from frequenz.channels import Broadcast, Receiver, Sender
 
-from ...actor import ChannelRegistry, ComponentMetricRequest
+from ...actor._channel_registry import ChannelRegistry
+from ...actor._data_sourcing.microgrid_api_source import ComponentMetricRequest
 from ...microgrid.component import ComponentMetricId
 from .._sample import Sample
 from ._formula_builder import FormulaBuilder
diff --git a/src/frequenz/sdk/timeseries/logical_meter/_tokenizer.py b/src/frequenz/sdk/timeseries/_logical_meter/_tokenizer.py
similarity index 100%
rename from src/frequenz/sdk/timeseries/logical_meter/_tokenizer.py
rename to src/frequenz/sdk/timeseries/_logical_meter/_tokenizer.py
```

</details>

Reply:
Except downstream users will also have to import `from timeseries._sample import Sample` in their actors.

Reply:
> Except downstream users will also have to import from timeseries._sample import Sample in their actors. Or maybe not.

I don't think so.

> There is one more option which I like better: move 'LogicalMeter' outside, into a data_pipeline module. LogicalMeter is not a timeseries primitive, although it does produce timeseries, just like some of the actors. BatteryPool, etc can go inside data_pipeline as well.

For me it is in the realm of timeseries, as it takes timeseries as input (+ a formula explaining how to combine those timeseries) and produces a timeseries as output.

Reply:
```python
>>> from frequenz.sdk.timeseries import LogicalMeter
>>> from frequenz.sdk.actor import DataSourcingActor
>>> 
```

It works because no we resolved the cyclic imports internally, as long as the user doesn't introduce more cycles themselves, the SDK shouldn't introduce any.

But again for the future, I would avoid having cyclic sub-package dependencies altogether. And maybe a way is to move stuff in `timeseries` that is using the `actor` package somewhere else, but I would leave that restructuring for the future, when we have more stuff so we can picture a better structure instead of doing futurology now.

Reply:
thanks for the patch! done

Reply:
looks like it is still broken though.  luckily I created a separate commit.  Maybe move `Sample` to near `ComponentMetricRequest`?

Reply:
Is the patch that you provided the final version that works for you?  Or were there changes after you pasted it here?

Reply:
@leandro-lucarella-frequenz I've created a separate PR with your patch and removed it from this PR,  so that this can be taken in,  and we can look into the paths next week.

Reply:
Ah, sorry, I didn't run the whole tests, I only did the quick import check. It looks like you are right, because Python needs to initialize the package `actor` to import, for example `actor._data_sourcing...` so it still runs into cyclic dependencies in some cases.

I guess there is no way around to avoid cyclic dependencies between sub-packages then, which will require us to shuffle things around a bit and clearly separate packages based on what they depend on, or have a more nested public interface where parent packages don't expose symbols of sub-packages.

Anyway, definitely something requiring more discussion and not for this PR.

Reply:
#84.